File: solutions/design/146-lru-cache.py
"""
# 146. LRU Cache
**Medium**

Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.

Implement the LRUCache class:
- LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
- int get(int key) Return the value of the key if the key exists, otherwise return -1.
- void put(int key, int value) Update the value of the key if it exists. Otherwise, add the key-value pair to the cache. If the number of keys exceeds the capacity from this operation, evict the least recently used key.

The functions get and put must each run in O(1) average time complexity.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
LRU Cache needs O(1) access and O(1) eviction. We can combine:
- HashMap for O(1) key lookup
- Doubly Linked List for O(1) insertion/deletion at any position

### APPROACH:
1. **HashMap + Doubly Linked List**: Hash map stores key->node mapping
2. **Recent Access Tracking**: Move accessed nodes to head of list
3. **Eviction Policy**: Remove from tail when capacity exceeded
4. **O(1) Operations**: All operations use direct node manipulation

### WHY THIS WORKS:
- HashMap provides O(1) key lookup
- Doubly linked list allows O(1) insertion/deletion anywhere
- Head represents most recently used, tail represents least recently used
- Moving nodes to head maintains LRU order efficiently

### TIME COMPLEXITY: O(1)
All operations (get, put) are constant time

### SPACE COMPLEXITY: O(capacity)
We store at most 'capacity' key-value pairs

### EXAMPLE WALKTHROUGH:
```
LRUCache(2)
put(1, 1) -> [1:1]
put(2, 2) -> [2:2, 1:1]
get(1)    -> [1:1, 2:2] (1 moves to front)
put(3, 3) -> [3:3, 1:1] (2 evicted)
get(2)    -> -1 (not found)
```

### EDGE CASES:
- Capacity of 1
- Get non-existent key
- Put same key multiple times
- Cache at full capacity

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

HashMap + Doubly Linked List approach:
1. Create DoublyLinkedNode class
2. Maintain head and tail dummy nodes
3. HashMap maps key to node
4. Get: return value and move to head
5. Put: add new or update existing, move to head, evict if needed

### Algorithm Steps:
1. Initialize head/tail dummy nodes and HashMap
2. Get: check HashMap, move node to head if found
3. Put: update existing or create new, move to head, evict tail if over capacity
4. Helper methods: addToHead, removeNode, moveToHead, removeTail

</details>
"""

class DoublyLinkedNode:
    """Node for doubly linked list."""
    def __init__(self, key=0, value=0):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        """
        Initialize LRU Cache with given capacity.

        Args:
            capacity: Maximum number of key-value pairs to store
        """
        self.capacity = capacity
        self.cache = {}  # key -> node mapping

        # Create dummy head and tail nodes
        self.head = DoublyLinkedNode()
        self.tail = DoublyLinkedNode()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _add_to_head(self, node: DoublyLinkedNode) -> None:
        """Add node right after head."""
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def _remove_node(self, node: DoublyLinkedNode) -> None:
        """Remove node from linked list."""
        node.prev.next = node.next
        node.next.prev = node.prev

    def _move_to_head(self, node: DoublyLinkedNode) -> None:
        """Move node to head (mark as most recently used)."""
        self._remove_node(node)
        self._add_to_head(node)

    def _remove_tail(self) -> DoublyLinkedNode:
        """Remove and return the last node (least recently used)."""
        last_node = self.tail.prev
        self._remove_node(last_node)
        return last_node

    def get(self, key: int) -> int:
        """
        Get value by key and mark as recently used.

        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        if key in self.cache:
            node = self.cache[key]
            # Move to head (mark as recently used)
            self._move_to_head(node)
            return node.value
        return -1

    def put(self, key: int, value: int) -> None:
        """
        Put key-value pair into cache.

        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        if key in self.cache:
            # Update existing key
            node = self.cache[key]
            node.value = value
            self._move_to_head(node)
        else:
            # Add new key
            new_node = DoublyLinkedNode(key, value)

            if len(self.cache) >= self.capacity:
                # Remove least recently used
                tail_node = self._remove_tail()
                del self.cache[tail_node.key]

            # Add new node
            self.cache[key] = new_node
            self._add_to_head(new_node)

# Alternative implementation using OrderedDict (simpler but less educational)
from collections import OrderedDict

class LRUCacheOrderedDict:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = OrderedDict()

    def get(self, key: int) -> int:
        if key in self.cache:
            # Move to end (most recent)
            self.cache.move_to_end(key)
            return self.cache[key]
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            # Update and move to end
            self.cache[key] = value
            self.cache.move_to_end(key)
        else:
            # Add new key
            if len(self.cache) >= self.capacity:
                # Remove least recent (first item)
                self.cache.popitem(last=False)
            self.cache[key] = value


def test_solution():
    """
    Test cases for 146. LRU Cache.
    """
    # Test case 1: Basic operations
    lru = LRUCache(2)

    lru.put(1, 1)
    lru.put(2, 2)
    assert lru.get(1) == 1, f"Expected 1, got {lru.get(1)}"

    lru.put(3, 3)  # evicts key 2
    assert lru.get(2) == -1, f"Expected -1, got {lru.get(2)}"

    lru.put(4, 4)  # evicts key 1
    assert lru.get(1) == -1, f"Expected -1, got {lru.get(1)}"
    assert lru.get(3) == 3, f"Expected 3, got {lru.get(3)}"
    assert lru.get(4) == 4, f"Expected 4, got {lru.get(4)}"

    # Test case 2: Capacity of 1
    lru2 = LRUCache(1)
    lru2.put(2, 1)
    assert lru2.get(2) == 1
    lru2.put(3, 2)  # evicts key 2
    assert lru2.get(2) == -1
    assert lru2.get(3) == 2

    # Test case 3: Update existing key
    lru3 = LRUCache(2)
    lru3.put(1, 1)
    lru3.put(2, 2)
    lru3.put(1, 10)  # update key 1
    assert lru3.get(1) == 10
    assert lru3.get(2) == 2

    # Test OrderedDict implementation
    lru_od = LRUCacheOrderedDict(2)
    lru_od.put(1, 1)
    lru_od.put(2, 2)
    assert lru_od.get(1) == 1
    lru_od.put(3, 3)  # evicts key 2
    assert lru_od.get(2) == -1

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    print("=== 146. LRU Cache ===")
    lru = LRUCache(2)

    print("put(1, 1)")
    lru.put(1, 1)
    print("put(2, 2)")
    lru.put(2, 2)
    print(f"get(1): {lru.get(1)}")
    print("put(3, 3)")  # evicts key 2
    lru.put(3, 3)
    print(f"get(2): {lru.get(2)}")  # returns -1 (not found)
    print(f"get(3): {lru.get(3)}")
    print(f"get(1): {lru.get(1)}")

File: docs/solutions/design/alternatives/146-lru-cache.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 * **Step 1:** [description]
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Node class for doubly-linked list
 */
class ListNode {
    constructor(key = 0, value = 0) {
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

/**
 * LRU Cache implementation
 *
 * @param {number} capacity - Maximum number of key-value pairs to store
 */
class LRUCache {
    constructor(capacity) {
        this.capacity = capacity;
        this.cache = new Map(); // key -> node mapping

        // Create dummy head and tail nodes for easier list manipulation
        this.head = new ListNode();
        this.tail = new ListNode();
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    /**
     * Get value for given key
     * @param {number} key
     * @return {number}
     */
    get(key) {
        const node = this.cache.get(key);
        if (!node) {
            return -1;
        }

        // Move accessed node to head (most recently used)
        this.moveToHead(node);
        return node.value;
    }

    /**
     * Put key-value pair in cache
     * @param {number} key
     * @param {number} value
     */
    put(key, value) {
        const existingNode = this.cache.get(key);

        if (existingNode) {
            // Update existing key
            existingNode.value = value;
            this.moveToHead(existingNode);
        } else {
            // Add new key
            const newNode = new ListNode(key, value);

            if (this.cache.size >= this.capacity) {
                // Remove least recently used (tail)
                const tail = this.removeTail();
                this.cache.delete(tail.key);
            }

            this.cache.set(key, newNode);
            this.addToHead(newNode);
        }
    }

    /**
     * Add node right after head
     * @param {ListNode} node
     */
    addToHead(node) {
        node.prev = this.head;
        node.next = this.head.next;
        this.head.next.prev = node;
        this.head.next = node;
    }

    /**
     * Remove node from its current position
     * @param {ListNode} node
     */
    removeNode(node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    /**
     * Move node to head (mark as most recently used)
     * @param {ListNode} node
     */
    moveToHead(node) {
        this.removeNode(node);
        this.addToHead(node);
    }

    /**
     * Remove and return tail node (least recently used)
     * @return {ListNode}
     */
    removeTail() {
        const lastNode = this.tail.prev;
        this.removeNode(lastNode);
        return lastNode;
    }
}

/**
 * Factory function for creating LRU Cache instances
 * @param {number} capacity
 * @return {LRUCache}
 */
function solve(capacity) {
    return new LRUCache(capacity);
}

/**
 * Test cases for Problem 146: LRU Cache
 */
function testSolution() {
    console.log('Testing 146. LRU Cache');

    // Test case 1: Basic functionality
    const lru1 = new LRUCache(2);
    lru1.put(1, 1);
    lru1.put(2, 2);
    console.assert(lru1.get(1) === 1, 'Test 1a failed');
    lru1.put(3, 3); // evicts key 2
    console.assert(lru1.get(2) === -1, 'Test 1b failed');
    lru1.put(4, 4); // evicts key 1
    console.assert(lru1.get(1) === -1, 'Test 1c failed');
    console.assert(lru1.get(3) === 3, 'Test 1d failed');
    console.assert(lru1.get(4) === 4, 'Test 1e failed');

    // Test case 2: Update existing key
    const lru2 = new LRUCache(2);
    lru2.put(1, 1);
    lru2.put(2, 2);
    lru2.put(1, 10); // update key 1
    console.assert(lru2.get(1) === 10, 'Test 2a failed');
    console.assert(lru2.get(2) === 2, 'Test 2b failed');

    // Test case 3: Capacity of 1
    const lru3 = new LRUCache(1);
    lru3.put(1, 1);
    console.assert(lru3.get(1) === 1, 'Test 3a failed');
    lru3.put(2, 2);
    console.assert(lru3.get(1) === -1, 'Test 3b failed');
    console.assert(lru3.get(2) === 2, 'Test 3c failed');

    // Test case 4: Access order affects eviction
    const lru4 = new LRUCache(2);
    lru4.put(1, 1);
    lru4.put(2, 2);
    lru4.get(1); // makes 1 most recently used
    lru4.put(3, 3); // should evict 2, not 1
    console.assert(lru4.get(1) === 1, 'Test 4a failed');
    console.assert(lru4.get(2) === -1, 'Test 4b failed');
    console.assert(lru4.get(3) === 3, 'Test 4c failed');

    console.log('All test cases passed for 146. LRU Cache!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 146. LRU Cache ===');
    console.log('Category: Design');
    console.log('Difficulty: Medium');
    console.log('');

    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    LRUCache,
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution uses classic LRU Cache design pattern
 * - Doubly-linked list + HashMap is the standard approach for O(1) operations
 * - Critical for system design interviews and real-world caching systems
 * - The dummy head/tail nodes simplify edge case handling
 */
